### PR TITLE
Egress Rules refactoring

### PR DESCRIPTION
**Pull request by vadimeisenbergibm**
**What this PR does / why we need it**:
Refactors handling egress rules to enable effective handling of multiple TCP and HTTP protocols.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
This PR is required to enable mongo and redis TCP protocols.

**Release note**:

```NONE
```

### CODE REVIEW

File: pilot/model/egress_rules.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+// IsEgressRulesSupportedProtocol returns true if the protocol is supported by egress rules
+func IsEgressRulesSupportedProtocol(protocol Protocol) bool {
+	return IsEgressRulesSupportedHTTPProtocol(protocol) || IsEgressRulesSupportedTCPProtocol(protocol)
+}
+
+func protocolMapAsString(protocolMap map[Protocol]bool) string {
+	first := true
+	result := ""
```
Comment (ijsnellf) [inline pilot/model/egress_rules.go]: maybe use `&bytes.Buffer{}` here to speed up the string concatenation?: https://golang.org/pkg/bytes/#Buffer

File: pilot/model/egress_rules.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import (
+	"fmt"
+	"sort"
+
+	multierror "github.com/hashicorp/go-multierror"
+	proxyconfig "istio.io/api/proxy/v1/config"
+)
+
+// RejectConflictingEgressRules rejects rules that have the destination which is equal to
+// the destionation of some other rule.
```
Comment (ijsnellf) [inline pilot/model/egress_rules.go]: destionation -> destination

File: pilot/model/validation.go [hunk 3]
```diff
@@ -877,9 +877,11 @@ func ValidateEgressRule(msg proto.Message) error {
 			errs = multierror.Append(errs, err)
 		}
 
-		if cidrDestinationService && Protocol(strings.ToUpper(port.Protocol)) != ProtocolTCP {
-			errs = multierror.Append(errs, fmt.Errorf("Only TCP protocol can be defined for CIDR destination "+
-				"service notation. port: %d protocol: %s destination.service: %s",
+		if cidrDestinationService &&
+			!IsEgressRulesSupportedTCPProtocol(Protocol(strings.ToUpper(port.Protocol))) {
+			errs = multierror.Append(errs, fmt.Errorf("Only the following protocols can be defined for "+
+				"CIDR destination service notation "+egressRulesSupportedTCPProtocols()+
```
Comment (ijsnellf) [inline pilot/model/validation.go]: nit: use string formatting to build the entire string instead of a mix of string concatenation and string formatting?

**Comment by vadimeisenbergibm** [comment]
Will rebase in a new branch and open as a separate PR.

**Pull request closed without merging**